### DISCUSSION TITLE
Prometheus Exporter with boost::beast Webserver

### DISCUSSION THREAD

**Discussion by dufferzafar** [open-telemetry/opentelemetry-cpp#2554] [Q&A] [answered]
I'm adding a bunch of metrics to an application that uses `boost::beast` based webserver. I'd like to have a `/metrics` endpoint, but WITHOUT using the current PrometheusExporter (and it's associated civet webserver).

Has anyone tried something similar? What changes would need to happen for this? 

I guess I'd still need a `PrometheusCollector` (and it's associated `TranslateToPrometheus` layer) then, replace `Prometheus::Exposer` with a function that returns a `String` in the prometheus metric format.

```
std::string prometheusMetricsView(std::weak_ptr<prometheus::Collectable> wcollectable)
{
    auto collectable = wcollectable.lock();
    if (!collectable) {
        return "";
    }

    std::ostringstream output;
    prometheus::TextSerializer serializer;
    auto&& metrics = collectable->Collect();
    serializer.Serialize(metrics, &output);
    return output.str();
}
```

I could then attach this function to a particular boost::beast view. 

Does this sound reasonable? Am I missing something obvious?

**Comment by dufferzafar** [comment]
One thing I missed in my original post is the link between an Exporter / MetricReader & a MeterProvider. So I still need a dummy exporter/reader that has a collector and can be passed into `MetricProvider::AddMetricReader()`

Not sure yet exactly how that interaction will look like.

**Comment by dufferzafar** [comment, marked as answer]
This is what I finally ended up with:

```cpp
#pragma once

#include <memory>
#include <mutex>
#include <sstream>
#include <string>
#include <vector>

#include <opentelemetry/exporters/prometheus/exporter_utils.h>
#include <opentelemetry/sdk/metrics/metric_reader.h>
#include <prometheus/collectable.h>
#include <prometheus/text_serializer.h>

class PrometheusExporter : public opentelemetry::sdk::metrics::MetricReader
{
public:
    opentelemetry::sdk::metrics::AggregationTemporality GetAggregationTemporality(
        opentelemetry::sdk::metrics::InstrumentType /*instrument_type*/) const noexcept override
    {
        return opentelemetry::sdk::metrics::AggregationTemporality::kCumulative;
    };

    std::vector<prometheus::MetricFamily> CollectPrometheus()
    {
        if (IsShutdown())
        {
            return {};
        }

        collection_lock_.lock();
        std::vector<prometheus::MetricFamily> result;
        auto status = Collect([&result](opentelemetry::sdk::metrics::ResourceMetrics& metric_data) {
            auto prometheus_metric_data
                = opentelemetry::exporter::metrics::PrometheusExporterUtils::TranslateToPrometheus(
                    metric_data, true, true);
            for (auto& data : prometheus_metric_data)
                result.emplace_back(data);
            return true;
        });
        collection_lock_.unlock();
        return result;
    }

    std::string serialize()
    {
        // TODO: Do we need to allocate this stream here?
        // What if we took the boost stream itself as an arg & wrote to it?
        std::ostringstream output;
        prometheus::TextSerializer().Serialize(output, CollectPrometheus());
        return output.str();
    };

private:
    mutable std::mutex collection_lock_;

    bool OnForceFlush(std::chrono::microseconds /*timeout*/) noexcept override { return true; };
    bool OnShutDown(std::chrono::microseconds /*timeout*/) noexcept override { return true; };
};
```

**Reply by dufferzafar** [reply]
Note that this also requires a separate target if you want to avoid linking against civetweb: #2565

**Comment by mohnishkodnani** [comment]
I have a similar use case where, i don't need an endpoint exposed through the prometheus exporter since I have my own http server. But on this endpoint i do want metrics in prometheus output format. 
How do you hook this custom MetricReader into the framework. Also, i was going over the API docs and could we use [opentelemetry-cpp](https://github.com/open-telemetry/opentelemetry-cpp/tree/main)/[exporters](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/exporters)/[ostream](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/exporters/ostream)/[src](https://github.com/open-telemetry/opentelemetry-cpp/tree/main/exporters/ostream/src)
/metric_exporter.cc
And may override the PushMetricExporter interface. I can't find in the documents around the MetricReader vs MetricExporter

**Reply by mohnishkodnani** [reply]
Also, with this method, I am not seeing the view take effect
